cmd/createtree/main.go

```diff
@@ -142,84 +126,32 @@ func newRequest(opts *createOpts) (*trillian.CreateTreeRequest, error) {
 		HashStrategy:       trillian.HashStrategy(hs),
 		HashAlgorithm:      sigpb.DigitallySigned_HashAlgorithm(ha),
 		SignatureAlgorithm: sigpb.DigitallySigned_SignatureAlgorithm(sa),
-		DisplayName:        opts.displayName,
-		Description:        opts.description,
+		DisplayName:        *displayName,
+		Description:        *description,
 		PrivateKey:         pk,
-		MaxRootDuration:    ptypes.DurationProto(opts.maxRootDuration),
+		MaxRootDuration:    ptypes.DurationProto(*maxRootDuration),
 	}}
 	return ctr, nil
 }
 
-func newPK(opts *createOpts) (*any.Any, error) {
-	switch opts.privateKeyType {
-	case "PEMKeyFile":
-		if opts.pemKeyPath == "" {
-			return nil, errors.New("empty pem_key_path")
-		}
-		if opts.pemKeyPass == "" {
-			return nil, fmt.Errorf("empty password for PEM key file %q", opts.pemKeyPath)
-		}
-		pemKey := &keyspb.PEMKeyFile{
-			Path:     opts.pemKeyPath,
-			Password: opts.pemKeyPass,
-		}
-		return ptypes.MarshalAny(pemKey)
-	case "PrivateKey":
-		if opts.pemKeyPath == "" {
-			return nil, errors.New("empty pem_key_path")
-		}
-		pemSigner, err := keys.NewFromPrivatePEMFile(
-			opts.pemKeyPath, opts.pemKeyPass)
+func newPK(format string) (*any.Any, error) {
+	if handler, ok := keyHandlers[format]; ok {
+		pb, err := handler()
 		if err != nil {
 			return nil, err
 		}
-		der, err := keys.MarshalPrivateKey(pemSigner)
-		if err != nil {
-			return nil, err
-		}
-		return ptypes.MarshalAny(&keyspb.PrivateKey{Der: der})
-	case "PKCS11ConfigFile":
-		if opts.pkcs11ConfigPath == "" {
-			return nil, errors.New("empty PKCS11 config file path")
-		}
-		configBytes, err := ioutil.ReadFile(opts.pkcs11ConfigPath)
-		if err != nil {
-			return nil, err
-		}
-		var config pkcs11key.Config
-		if err = json.Unmarshal(configBytes, &config); err != nil {
-			return nil, err
-		}
-		pubKeyBytes, err := ioutil.ReadFile(config.PublicKeyPath)
-		if err != nil {
-			return nil, err
-		}
-		return ptypes.MarshalAny(&keyspb.PKCS11Config{
-			TokenLabel: config.TokenLabel,
-			Pin:        config.PIN,
-			PublicKey:  string(pubKeyBytes),
-		})
-	default:
-		return nil, fmt.Errorf("unknown private key type: %v", opts.privateKeyType)
+		return ptypes.MarshalAny(pb)
 	}
+
+	return nil, fmt.Errorf("private key format must be one of: %s", strings.Join(keyFormats(), ", "))
 }
 
-func newOptsFromFlags() *createOpts {
-	return &createOpts{
-		addr:             *adminServerAddr,
-		treeState:        *treeState,
-		treeType:         *treeType,
-		hashStrategy:     *hashStrategy,
-		hashAlgorithm:    *hashAlgorithm,
-		sigAlgorithm:     *signatureAlgorithm,
-		displayName:      *displayName,
-		description:      *description,
-		maxRootDuration:  *maxRootDuration,
-		privateKeyType:   *privateKeyFormat,
-		pemKeyPath:       *pemKeyPath,
-		pemKeyPass:       *pemKeyPassword,
-		pkcs11ConfigPath: *pkcs11ConfigPath,
+func keyFormats() []string {
```

Comment:
Done (though it's now called `RegisteredTypes()` in the createtree/keys package).